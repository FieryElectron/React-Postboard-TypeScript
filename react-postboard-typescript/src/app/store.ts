import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';

import signInPageReducer from '../components/SignInPage/signInPageSlice';
import postBoardPageReducer from '../components/PostBoardPage/postBoardPageSlice';

export const store = configureStore({
  reducer: {
    signInPage: signInPageReducer,
    postBoardPage: postBoardPageReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
