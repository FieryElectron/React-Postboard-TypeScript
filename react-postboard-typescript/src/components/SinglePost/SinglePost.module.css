.singlePostContainer {
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
    transition: 0.3s;
    margin: 5px;
    padding: 5px;
    border-radius: 5px;
}

.ownerName {
    text-align: left;
    left: 1%;
    position: relative;
    font-size: 25px;
}

.postContent {
    text-align: left;
    left: 1%;
    position: relative;
    font-weight: 400;
    font-size: 15px;
}

.timeStamp {
    padding: 10px;
    font-weight: 400;
    font-size: 15px;
    height: 32px;
}

.timeStampContainer {
    display: flex;
    justify-content: space-between;
    align-items: center;
}