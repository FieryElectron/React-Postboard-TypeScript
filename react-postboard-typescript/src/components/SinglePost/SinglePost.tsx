import styles from './SinglePost.module.css';

import Button from '../Button/Button';
import { useNavigate } from 'react-router-dom';


import { useAppSelector, useAppDispatch } from '../../app/hooks';

import axios from 'axios';
import { toast } from 'react-toastify';

import {
    selectToken,
    selectUsername,
    selectPost,
    getPosts,
} from '../PostBoardPage/postBoardPageSlice';


const refreshAccessToken = async (tokenUrl : string) => {
    const res = await axios.get(tokenUrl, {withCredentials: true});
    return res.data.flag;
}




const SinglePost = ( post : any) => {
    const dispatch = useAppDispatch();
    const navigate = useNavigate();

    const currentUsername = useAppSelector(selectUsername);
    const tokenUrl = useAppSelector(selectToken);
    const postUrl = useAppSelector(selectPost);


    const deletePost = async () => {
        const url = postUrl+`${post.post.id}`;
        const res = await axios.delete(url, {withCredentials: true});

        if (res.data.flag) {
            toast.info(res.data.info, {
                position: 'top-center',
                autoClose: 1000,
                pauseOnFocusLoss:false,
            });

            dispatch(getPosts(navigate));
            return;
        }

        if(res.data.info){
            toast.info(res.data.info, {
                position: 'top-center',
                autoClose: 1000,
                pauseOnFocusLoss:false,
            });
        }


        if (res.data.refreshToken) {
            if (await refreshAccessToken(tokenUrl)) {
                const resSecond = await axios.delete(url, {withCredentials: true});
                if (resSecond.data.flag) {
                    toast.info(res.data.info, {
                        position: 'top-center',
                        autoClose: 1000,
                        pauseOnFocusLoss:false,
                    });
        
                    dispatch(getPosts(navigate));
                    return;
                }
            }
        }
    
        // navigate('/');
        return;
    }

    return (
        <div data-testid='singlePostContainer' className={styles.singlePostContainer}>
            <div data-testid='ownerName'  className={styles.ownerName}>{post.post.username}</div>
            <div data-testid='postContent'  className={styles.postContent}>{post.post.content}</div>
            <hr></hr>

            <div data-testid='timeStampContainer' className={styles.timeStampContainer}>
                <span data-testid='timeStamp' className={styles.timeStamp}>{new Date(post.post.timestamp).toLocaleDateString('en-US')}</span>
                {currentUsername === post.post.username && <Button text='Delete' color='#E62E2E' onClick={deletePost}></Button>}
                
            </div>
        </div>
    )
}

SinglePost.defaultProps = {
    ownerName: 'ownerName',
    Content: 'Content',
    timestamp: 'timestamp',
}

export default SinglePost;