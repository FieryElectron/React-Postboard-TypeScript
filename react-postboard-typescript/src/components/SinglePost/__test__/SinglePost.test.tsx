import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom/extend-expect';

import SinglePost from '../SinglePost';

import { BrowserRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux';
import { store } from '../../../app/store';

let container: HTMLElement;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container.remove();
});

it('empty test', () => {

});

it('renders SinglePost without crashing', () => {
  const post = {
    post: {
      username: 'username',
      content: 'content',
      timestamp: 'timestamp',
    }
  }

  act(() => {
    render(<Provider store={store}><Router><SinglePost post={post}/></Router></Provider>);
  });
});

it('renders SinglePost class correctly', () => {
  const post = {
    post: {
      username: 'username',
      content: 'content',
      timestamp: 'timestamp',
    }
  }

  act(() => {
    render(<Provider store={store}><Router><SinglePost post={post}/></Router></Provider>);
  });

  const singlePostContainer = document.querySelector('[data-testid=singlePostContainer]')!;
  const ownerName = document.querySelector('[data-testid=ownerName]')!;
  const postContent = document.querySelector('[data-testid=postContent]')!;
  const timeStampContainer = document.querySelector('[data-testid=timeStampContainer]')!;
  const timeStamp = document.querySelector('[data-testid=timeStamp]')!;



  expect(singlePostContainer).toHaveAttribute('class', 'singlePostContainer');
  expect(ownerName).toHaveAttribute('class', 'ownerName');
  expect(postContent).toHaveAttribute('class', 'postContent');
  expect(timeStampContainer).toHaveAttribute('class', 'timeStampContainer');
  expect(timeStamp).toHaveAttribute('class', 'timeStamp');
});

it('matchs snapshot 1', ()=>{
  const post = {
    post: {
      username: 'username',
      content: 'content',
      timestamp: 'timestamp',
    }
  }

  const tree = renderer.create(<Provider store={store}><Router><SinglePost post={post} /></Router></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});