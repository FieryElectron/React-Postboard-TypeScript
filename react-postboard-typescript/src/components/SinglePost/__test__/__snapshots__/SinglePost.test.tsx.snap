// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchs snapshot 1 1`] = `
<div
  className="singlePostContainer"
  data-testid="singlePostContainer"
>
  <div
    className="ownerName"
    data-testid="ownerName"
  />
  <div
    className="postContent"
    data-testid="postContent"
  />
  <hr />
  <div
    className="timeStampContainer"
    data-testid="timeStampContainer"
  >
    <span
      className="timeStamp"
      data-testid="timeStamp"
    >
      Invalid Date
    </span>
  </div>
</div>
`;
